library/src/main/java/com/davidstemmer/screenplay/flow/Screenplay.java

```diff
@@ -14,73 +14,94 @@
+                    new ArrayDeque<Scene>();
         }
         else {
-            // Using forward/replace, there is always one (and only one) incoming scene
-            incomingScenes = new ArrayDeque<>();
-            incomingScenes.push(incomingScene);
-            // Forward/replace to a stacked scene:
-            // No outgoing scenes to animate or tear down
-            if (incomingScene.isStacking() || previousScenes.isEmpty()) {
-                outgoingScenes = new ArrayDeque<>();
-            }
-            // Forward/replace to a non-stacked scene:
-            // Animate and tear down the previous scene block
-            else {
-                outgoingScenes = previousScenes;
-            }
+            ArrayDeque<Scene> difference = difference(nextBackstack, previousBackstack);
+
+            animatedScenesIn = getLastSceneBlock(difference);
+            animatedScenesOut = moveToNewSceneBlock(difference) ?
+                    getLastSceneBlock(previousBackstack) :
+                    new ArrayDeque<Scene>();
 
         }
 
         Scene.Transformer delegatedTransformer = direction == Flow.Direction.BACKWARD ?
-                outgoingScenes.getFirst().getTransformer():
-                incomingScene.getTransformer();
+                animatedScenesOut.iterator().next().getTransformer():
+                animatedScenesIn.iterator().next().getTransformer();
+
+        sceneCut.setIncomingScenes(animatedScenesIn);
+        sceneCut.setOutgoingScenes(animatedScenesOut);
 
-        for (Scene scene : incomingScenes) {
-            sceneCut.addIncomingScene(scene);
+        previousBackstack.clear();
+        previousBackstack.addAll(nextBackstack);
+
+        beginCut(sceneCut.build(), delegatedTransformer);
+    }
+
+    private ArrayDeque<Scene> deque(Backstack backstack) {
+        ArrayDeque<Scene> sceneSet = new ArrayDeque<Scene>();
+        for (Backstack.Entry entry : backstack) {
+            Scene nextScene = (Scene) entry.getScreen();
+            sceneSet.add(nextScene);
         }
+        return sceneSet;
+    }
 
-        for (Scene scene : outgoingScenes) {
-            sceneCut.addOutgoingScene(scene);
+    private boolean moveToNewSceneBlock(ArrayDeque<Scene> difference) {
+        for (Scene scene : difference) {
+            if (!scene.isStacking()) {
+                return true;
+            }
         }
+        return false;
+    }
 
-        previousScenes.clear();
-        previousScenes.addAll(getLastSceneBlock(nextBackstack));
+    private ArrayDeque<Scene> difference(ArrayDeque<Scene> incoming,
+                                         ArrayDeque<Scene> outgoing) {
+        ArrayDeque<Scene> difference = new ArrayDeque<>(incoming);
+        difference.removeAll(outgoing);
+        return validateDifference(difference);
+    }
 
-        beginCut(sceneCut.build(), delegatedTransformer);
+    private ArrayDeque<Scene> validateDifference(ArrayDeque<Scene> difference) {
```

Comment:
What's the name of this pattern, where you pass something in, inspect it, and then return it unchanged (obviously throwing doesn't return it)? Just curious.